Explain how to close an issue within a commit message

asalkeld opened the pull request:
## Pull Request Guidelines

This adds some guidance on how to close issues automatically.

thockin commented on .github/PULL_REQUEST_TEMPLATE.md:
````diff
@@ -11,5 +11,11 @@
 ```
 
+If you want this PR to close an issue add the following:
+"fixes #<issue nubmber>"
````
this should be `backtick quoted` or else the `<issue nubmber>` gets eaten as HTML

"nubmber" typo

thockin commented on .github/PULL_REQUEST_TEMPLATE.md:
````diff
@@ -11,5 +11,11 @@
 * Enter your extended release note here
 ```
 
+If you want this PR to close an issue add the following:
````
s/close/automatically close

thockin commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -11,5 +11,11 @@
 
+If you want this PR to close an issue add the following:
+"fixes #<issue nubmber>"
+either in this pull request message or the commit message.
```
Prefer not to encourage embedding PR #s in commit messages.  Can you reword as "If you want this PR to automatically close an issue when it is merged, add the following to the pull request message:".  Maybe also note that a PR can close multiple issues.

thockin commented:
Can you also change the title line to `## Checklist for submitting a Pull Request (delete this when complete)` or something like that?  Many people forget to delete some or all of it.

asalkeld commented:
Thanks, new PR version on the way...

bgrant0607 commented:
cc @david-mcmahon

bgrant0607 commented:
Thanks!

As long as you're doing this, could you ask people to delete the template, too (except for the release note, if they added one)? We've noticed a lot of PRs don't touch it. For instance: #25746

david-mcmahon commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,15 +1,26 @@
+
+`fixes #<issue number>, fixes #<issue number>, fixes #<issue number>`
+
+See: https://github.com/blog/1506-closing-issues-via-pull-requests
```
Why not link the text "want this PR to automatically close an issue" to save verbosity on this template.  The less people have to read, the more likely they will.

david-mcmahon commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,15 +1,26 @@
-## Pull Request Guidelines
+## Checklist for submitting a Pull Request (delete this when complete)
```
"delete this when complete" is not what we want here. 
If people use this template to add a more detailed release note, at least that block needs to remain.  
In fact, that is the primary reason for this PR template so if nothing else, that should be highlighted.

Commit:
Explain how to close an issue within a PR message

- updated the headline
- updated the checklist numbers (was all "1.")
- converted the instructions into a markdown comment to make it
  clearer what is an instruction.

asalkeld commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,8 +1,15 @@
-## Pull Request Guidelines
+<!--
```
I have tried to please everyone by making it more obvious what is an instruction (in the markdown comment) and saying below that the author must delete the comment.

david-mcmahon commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,8 +1,15 @@
+<!--
+Checklist for submitting a Pull Request
+
+Please remove this comment block before submitting.
```
How about "(If an expanded release note is not needed, this entire block may be deleted)"

Also you've got <!-- comments around the entire list.

asalkeld commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,8 +1,15 @@
+   add `fixes #<issue number>` or `fixes #<issue number>, fixes #<issue number>`
+   to close multiple issues (see: https://github.com/blog/1506-closing-issues-via-pull-requests).
+4. Follow the instructions for [labeling and writing a release note for this PR](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes) in the block below.
+-->
```
Hi @david-mcmahon , yes I comment out the whole list of instructions because you don't want any of them in your final PR. I am trying to show the developer (a bit like git commit) that what is in a comment is "for you info only" and what you need to leave is just the release-note.

david-mcmahon replied:
Oh I see, that was intentional.  Ultimately this leaves the important block (with instructions) in there, but if the developer hurries past the initial PR screen the commented block will be invisible otherwise - unless the developer presses "Edit" on that comment.  That feels a bit odd, but as I write this I'm starting to like it.  Yes, I like it.  Let's see how it works for people.

asalkeld commented on .github/PULL_REQUEST_TEMPLATE.md:
````diff
@@ -1,8 +1,15 @@
+-->
 
 ```release-note
 * Use the release-note-* labels to set the release note state 
````
Note: the release note below still has "Clear this block to use the PR title as the release note "

xiangpengzhao commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,8 +1,15 @@
-1. Follow the instructions for [labeling and writing a release note for this PR](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes) in the block below.
+2. See our [developer guide](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md).
+3. If you want this PR to automatically close an issue when it is merged,
+   add `fixes #<issue number>` or `fixes #<issue number>, fixes #<issue number>`
```
What if someone accidentally type an unintended issue number? Thanks! @asalkeld

thockin replied:
I cede any control I had to Mr. McMahon

On Tue, May 17, 2016 at 7:17 PM, Xiangpeng Zhao <notifications@github.com>
wrote:

> In .github/PULL_REQUEST_TEMPLATE.md
> <https://github.com/kubernetes/kubernetes/pull/25717#discussion_r63636299>
> :
>
> >
> >  1. Please read our [contributor guidelines](https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md).
> > -1. See our [developer guide](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md).
> > -1. Follow the instructions for [labeling and writing a release note for this PR](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes) in the block below.
> > +2. See our [developer guide](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md).
> > +3. If you want this PR to automatically close an issue when it is merged,
> > +   add `fixes #<issue number>` or `fixes #<issue number>, fixes #<issue number>`
>
> What if someone accidentally type an unintended issue number? Thanks!
> @asalkeld <https://github.com/asalkeld>
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25717/files/448b1d2fde7bee29e49ffbd882e3a01e0b40971d#r63636299>
>

asalkeld replied:
@xiangpengzhao I am not changing the mechanism (lots of developers are already using this github feature), I am just documenting it. To answer your question it will close the issue that you do place in the message, so if it's wrong - the wrong issue gets closed (hopefully you notice and fix it manually).

xiangpengzhao replied:
I got it. :smile:  Thanks! @thockin @asalkeld
